libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,575 @@
+   constexpr static bool valid = true;
+   bool operator()(const flat_set<FieldElement>& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      return std::includes(f.begin(), f.end(), a.begin(), a.end());
+   }
+};
+template<typename FieldContainer, typename ArgumentElement>
+struct predicate_has_all<FieldContainer, flat_set<ArgumentElement>,
+                         std::enable_if_t<is_container<FieldContainer> && !is_flat_set<FieldContainer> &&
+                                          comparable_types<typename FieldContainer::value_type, ArgumentElement>>> {
+   predicate_has_all<flat_set<typename FieldContainer::value_type>, flat_set<ArgumentElement>> inner;
+   // Field is other container; convert to flat_set
+   constexpr static bool valid = true;
+   bool operator()(const FieldContainer& f, const flat_set<ArgumentElement>& a) const {
+      if (f.size() < a.size()) return false;
+      flat_set<typename FieldContainer::value_type> fs(f.begin(), f.end());
+      return inner(fs, a);
+   }
+};
+template<typename OptionalType, typename Argument>
+struct predicate_has_all<fc::optional<OptionalType>, Argument, void> : predicate_has_all<OptionalType, Argument> {
+   // Field is optional container
+   bool operator()(const fc::optional<OptionalType>& f, const Argument& a) const {
+      if (!f.valid()) return predicate_result::Rejection(predicate_result::null_optional);
+      return (*this)(*f, a);
+   }
+};
+
+// List contains none of list predicate
+template<typename C1, typename C2, typename = void> struct predicate_has_none : predicate_invalid<C1, C2> {};
+template<typename FieldElement, typename ArgumentElement>
+struct predicate_has_none<flat_set<FieldElement>, flat_set<ArgumentElement>,
+                          std::enable_if_t<comparable_types<FieldElement, ArgumentElement>>> {
+   // Field is already flat_set
+   constexpr static bool valid = true;
+   bool operator()(const flat_set<FieldElement>& f, const flat_set<ArgumentElement>& a) const {
+      flat_set<FieldElement> intersection;
+      std::set_intersection(f.begin(), f.end(), a.begin(), a.end(),
+                            std::inserter(intersection, intersection.begin()));
+      return intersection.empty();
+   }
+};
+template<typename FieldContainer, typename ArgumentElement>
+struct predicate_has_none<FieldContainer, flat_set<ArgumentElement>,
+                          std::enable_if_t<is_container<FieldContainer> && !is_flat_set<FieldContainer> &&
+                                           comparable_types<typename FieldContainer::value_type, ArgumentElement>>> {
+   predicate_has_none<flat_set<typename FieldContainer::value_type>, flat_set<ArgumentElement>> inner;
+   // Field is other container; convert to flat_set
+   constexpr static bool valid = true;
+   bool operator()(const FieldContainer& f, const flat_set<ArgumentElement>& a) const {
+      flat_set<typename FieldContainer::value_type> fs(f.begin(), f.end());
```

Comment:
Constructing a `flat_set` can have `O(n^2)` when inserting in the wrong order.
Are there any operations that have containers with a possibly wrong sorting order?

Reply:
We have some operations with `vector<char>` which may not be critical because there can be only 255 different elements.

We also have some operations with more complex `vector`s, specifically in `confidential.hpp` and `proposal_create_operation::proposed_ops`. I'm unsure how these are handled because it seems the contains type is not sortable and could thus not be transformed into a `flat_set`. @MichelSantos do your additional tests cover such cases?

Reply:
Not yet.  Do you have any suggestions about the type of tests for the "has" and "in" predicates?

Reply:
Can you check if any of the collection-based predicates work with confidential inputs/outputs, and on `proposed_ops`?
Also check an operation with a long (several 1000) `vector<char>` against a long list of `and'd` `has_all` predicates, all of which are true.
Please measure runtime behaviour depending on vector length and number of predicates (this perhaps in `performance_test` suite, not in `chain_test`).

Reply:
If I use the `std::any_of` implementation for the `flat_set` specialization above, I no longer need `f` to be an ordered container anyways, so I could just use that same specialization for all containers and not just flat sets.

Reply:
Sounds good.

Reply:
So keep the `any_of` implementation here, since the vector can't be assumed sorted, but keep the `set_intersection` specialization for `flat_set` fields?

Reply:
`std::any_of` here is good.

Reply:
> Can you check if any of the collection-based predicates work with confidential inputs/outputs, and on `proposed_ops`?

The current implementation/spec shouldn't be able to do much with those field types... None of `has_all`, `has_none`, `in`, and `not_in` should work -- all should throw an exception when creating the predicate.